src/Interpreters/HashJoin.cpp

```diff
@@ -1826,6 +1963,15 @@ void HashJoin::checkTypesOfKeys(const Block & block) const
 }
 
 void HashJoin::joinBlock(Block & block, ExtraBlockPtr & not_processed)
+{
+    auto stream_output = joinBlockWithStreamOutput(block, not_processed);
+    StreamReplicateBlocks * stream_replicate_blocks = static_cast<StreamReplicateBlocks *>(stream_output.get());
+    stream_replicate_blocks->disableStream();
```

Comment:
Just wondering is there cases when after setting `disableStream` it works faster that reading and concatenating all the blocks? Don't we need a dyncamic dispatch here based on some heuristic to choose if we need to stream or just use current code

Reply:
Whether streaming output is required here depends on the data. When there is some unbalanced data or the order header can match a lot of order lines, it will cause the memory to be out of control. There is no way to find a fixed rule to determine whether the current join will produce huge output.

I've tried reading and concatenating all the blocks in concurrent hash. The overhead introduced is enough to wipe out the optimization.